src/hotspot/share/runtime/signature.cpp

```diff
@@ -435,6 +435,30 @@ Symbol* Signature::strip_envelope(const Symbol* signature) {
                                  signature->utf8_length() - 2);
 }
 
+const char* Signature::field_type_string(const Symbol* signature) {
+  return field_type_string_impl(signature, 0);
+}
+
+const char* Signature::field_type_string_impl(const Symbol* signature, int index) {
+  if (basic_type(signature->char_at(index)) == T_OBJECT) {
+    return strip_envelope(signature)->as_C_string();
+  } else if (basic_type(signature->char_at(index)) == T_ARRAY) {
+    int i;
+    // Unravel and count array layers
+    for (i = 0; i < signature->utf8_length() && signature->char_at(i) == JVM_SIGNATURE_ARRAY; i++);
+    char* buf = NEW_RESOURCE_ARRAY(char, 64);
+    strncpy(buf, field_type_string_impl(signature, i), 64);
+    // Append array braces to type name
+    for (int j = 0; j < i; j++) {
+      strncat(buf, "[]", strlen(buf));
+      buf[strlen(buf)] = '\0';
+    }
```

Comment:
Java array dimensions can be up to 255, so the `buf` is not long enough.

Anyway, I think we already have code that can translate a signature like `[I` to the source code form of `int []`. Try this:

```
super class Foofoo
	version 63:0
{
  public static Method main:"([Ljava/lang/String;)V"
	stack 30 locals 3
  {
                iconst_0;
                iconst_0;
                aconst_null;
		invokestatic	Method "xxx":"(II[I)V";
		return;
  }
} 

$ java -cp . Foofoo
Exception in thread "main" java.lang.NoSuchMethodError: 'void Foofoo.xxx(int, int, int[])'
	at Foofoo.main(Foofoo.jasm)
```

Writing such conversion is very difficult. You should try to use the existing code (refactoring it as necessary).

Reply:
After a while of searching I was unable to find anything that could easily provide this type of information. The closest example is Signature::print_as_signature_external_return_type, but this does not work with fields.

Reply:
Ignore my above comment, I found a way to rework a method in symbol.cpp to accomplish the same task more safely. Thank you for this!